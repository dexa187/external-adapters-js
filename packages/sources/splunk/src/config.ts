import { util } from '@chainlink/ea-bootstrap'
import { RequestConfig } from '@chainlink/types'

export const NAME = 'SPLUNK' // This should be filled in with a name corresponding to the data provider using UPPERCASE and _underscores_.

export const DEFAULT_ENDPOINT = 'en-US/account/login'
export const DEFAULT_BASE_URL = 'https://localhost:8089'

export type Config = {
  host?: string
  port?: string
  user?: string
  password?: string
  scheme?: string

  api?: RequestConfig
}

export const makeConfig = (prefix?: string): Config => ({
  host: util.getRequiredEnv('HOST', prefix) || 'localhost',
  port: util.getRequiredEnv('PORT', prefix) || 8089,
  scheme: util.getRequiredEnv('SCHEME', prefix) || 'https',
  user: util.getRequiredEnv('USER', prefix) || 'admin',
  password: util.getRequiredEnv('PASSWORD', prefix),
  api: {},
})
