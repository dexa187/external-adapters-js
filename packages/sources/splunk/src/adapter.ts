import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { ExecuteWithConfig, ExecuteFactory } from '@chainlink/types'
import { Config, makeConfig } from './config'
var splunkjs = require('splunk-sdk')

const inputParams = {
  search: true,
}

export interface ResponseSchema {
  data: {
    result: any
  }
}

export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
  const validator = new Validator(request, inputParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const search = validator.validated.data.search

  //Make splunk search here
  const service = new splunkjs.Service({
    username: config.user,
    password: config.password,
    host: config.host,
    port: config.port,
    scheme: config.scheme,
  })

  const kwargs_blockingsearch = { exec_mode: 'blocking' }

  console.log(`${JSON.stringify(config)}`)

  const login = () => {
    return new Promise((resolve, reject) => {
      service.login((err: any, success: any) => {
        if (err != null) {
          console.log(err)
          reject(err)
        }
        resolve(success)
      })
    })
  }
  await login()
  const jobs = service.jobs()
  const searchJob = (search: string, kwargs_blockingsearch: any) => {
    return new Promise((resolve, reject) => {
      jobs.oneshotSearch(search, kwargs_blockingsearch, (err: any, results: any) => {
        if (err != null) {
          console.log(err)
          reject(err)
        }
        resolve(results)
      })
    })
  }

  const job: any = await searchJob(search, kwargs_blockingsearch)

  return Requester.success(jobRunID, { data: { result: `${job.rows[0]}` } }, true)
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
