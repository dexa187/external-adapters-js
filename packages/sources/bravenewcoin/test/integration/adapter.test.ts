import { AdapterRequest } from '@chainlink/types'
import http from 'http'
import nock from 'nock'
import request from 'supertest'
import { server as startServer } from '../../src'
import {
  mockAuthTokenResponse,
  mockBtcCoinEndpoint,
  mockBtcMarketEndpoint,
  mockCryptoResponse,
  mockEthCoinEndpoint,
  mockEthMarketEndpoint,
} from './cryptoFixtures'
import { mockEthOhlcvEndpoint, mockVwapResponse } from './vwapFixtures'

let oldEnv: NodeJS.ProcessEnv

describe('bravenewcoin', () => {
  let server: http.Server
  const req = request('localhost:8080')

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env.CACHE_ENABLED = 'false'
    process.env.API_KEY = process.env.API_KEY || 'mock-api-key'
    process.env.CLIENT_ID = process.env.CLIENT_ID || 'mock-client-id'
    if (process.env.RECORD) {
      nock.recorder.rec()
    }
    server = await startServer()
  })
  afterAll((done) => {
    process.env = oldEnv
    if (process.env.RECORD) {
      nock.recorder.play()
    }

    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
    server.close(done)
  })

  describe('when making a request to bravenewcoin to crypto endpoint', () => {
    const cryptoRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'crypto',
        base: 'ETH',
        quote: 'BTC',
      },
    }

    describe('bravenewcoin replies with success', () => {
      it('should reply with success', async () => {
        mockAuthTokenResponse()
        mockBtcCoinEndpoint()
        mockBtcMarketEndpoint()
        mockCryptoResponse()
        mockEthCoinEndpoint()
        mockEthMarketEndpoint()
        const response = await req
          .post('/')
          .send(cryptoRequest)
          .set('Accept', '*/*')
          .set('Content-Type', 'application/json')
          .expect('Content-Type', /json/)
          .expect(200)
        expect(response.body).toMatchSnapshot()
      })
    })
  })

  describe('when making a request to bravenewcoin to vwap endpoint', () => {
    const vwapRequest: AdapterRequest = {
      id: '2',
      data: {
        endpoint: 'vwap',
        base: 'ETH',
      },
    }

    describe('bravenewcoin replies with success', () => {
      it('should reply with success', async () => {
        mockAuthTokenResponse()
        mockEthCoinEndpoint()
        mockEthOhlcvEndpoint()
        mockVwapResponse()
        const response = await req
          .post('/')
          .send(vwapRequest)
          .set('Accept', '*/*')
          .set('Content-Type', 'application/json')
          .expect('Content-Type', /json/)
          .expect(200)
        expect(response.body).toMatchSnapshot()
      })
    })
  })
})
