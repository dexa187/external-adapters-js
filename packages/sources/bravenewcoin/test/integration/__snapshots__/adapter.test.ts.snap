// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bravenewcoin when making a request to bravenewcoin to crypto endpoint when sending request without API key should reply with failure 1`] = `
Object {
  "error": Object {
    "message": "Request failed with status code 403",
    "name": "AdapterError",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 403,
}
`;

exports[`bravenewcoin when making a request to bravenewcoin to crypto endpoint when sending well-formed request should reply with success 1`] = `
Object {
  "data": Object {
    "result": 0.05986385840449063,
  },
  "jobRunID": "1",
  "result": 0.05986385840449063,
  "statusCode": 200,
}
`;

exports[`bravenewcoin when making a request to bravenewcoin to vwap endpoint when sending request without API key should reply with failure 1`] = `
Object {
  "error": Object {
    "message": "Request failed with status code 403",
    "name": "AdapterError",
  },
  "jobRunID": "2",
  "status": "errored",
  "statusCode": 403,
}
`;

exports[`bravenewcoin when making a request to bravenewcoin to vwap endpoint when sending well-formed request should reply with success 1`] = `
Object {
  "data": Object {
    "result": 3821.520292860939,
  },
  "jobRunID": "2",
  "result": 3821.520292860939,
  "statusCode": 200,
}
`;
